# **Data Cleaning Exercise**

In [1]:
#All our necesary libraries and the data
import pandas as pd
import numpy as np

#read in the data
nfl_data = pd.read_csv("../input/nflplaybyplay2009to2016/NFL Play by Play 2009-2017 (v4).csv")

#set seed for reproducibility
np.random.seed(0) 

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3156: DtypeWarning: Columns (25,51) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
# Lets check the head of the data
# We can already see a few data points missing
nfl_data.head()

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
0  2009-09-10  2009091000      1    1   NaN  15:00         15    3600.0   
1  2009-09-10  2009091000      1    1   1.0  14:53         15    3593.0   
2  2009-09-10  2009091000      1    1   2.0  14:16         15    3556.0   
3  2009-09-10  2009091000      1    1   3.0  13:35         14    3515.0   
4  2009-09-10  2009091000      1    1   4.0  13:27         14    3507.0   

   PlayTimeDiff SideofField  ...    yacEPA  Home_WP_pre  Away_WP_pre  \
0           0.0         TEN  ...       NaN     0.485675     0.514325   
1           7.0         PIT  ...  1.146076     0.546433     0.453567   
2          37.0         PIT  ...       NaN     0.551088     0.448912   
3          41.0         PIT  ... -5.031425     0.510793     0.489207   
4           8.0         PIT  ...       NaN     0.461217     0.538783   

   Home_WP_post  Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.546433      0.453567  0.485675  0.060758       NaN       NaN    2009  
1      0.551088      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.510793      0.489207  0.551088 -0.040295       NaN       NaN    2009  
3      0.461217      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.558929      0.441071  0.461217  0.097712       NaN       NaN    2009  

[5 rows x 102 columns]

In [3]:
#Check the info on the data
nfl_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407688 entries, 0 to 407687
Columns: 102 entries, Date to Season
dtypes: float64(33), int64(31), object(38)
memory usage: 317.3+ MB


In [4]:
#How much data is missing
missing_vals = nfl_data.isnull().sum()

#Take a look at the first 15 columns
missing_vals[0:15]

Date                0
GameID              0
Drive               0
qtr                 0
down            61154
time              224
TimeUnder           0
TimeSecs          224
PlayTimeDiff      444
SideofField       528
yrdln             840
yrdline100        840
ydstogo             0
ydsnet              0
GoalToGo          840
dtype: int64

In [5]:
#What is the percentage missing cells

total_cells = np.product(nfl_data.shape)
total_missing = missing_vals.sum()

percent_missing = (total_missing/total_cells)*100
print ('{}%'.format(percent_missing))

24.87214126835169%


**That implies about a quarter of the data is missing. That is a lot**

In [6]:
#Several ways to go about handling our missing values
#One of the less recommended approaches but an easy quick approach nonetheless is removing all the unavailable data
#Take note to remove data accross columns not the rows
dropped_data = nfl_data.dropna(axis=1)
dropped_data.head()

Date      GameID  Drive  qtr  TimeUnder  ydstogo  ydsnet  \
0  2009-09-10  2009091000      1    1         15        0       0   
1  2009-09-10  2009091000      1    1         15       10       5   
2  2009-09-10  2009091000      1    1         15        5       2   
3  2009-09-10  2009091000      1    1         14        8       2   
4  2009-09-10  2009091000      1    1         14        8       2   

   PlayAttempted  Yards.Gained  sp  ...  Timeout_Indicator  Timeout_Team  \
0              1            39   0  ...                  0          None   
1              1             5   0  ...                  0          None   
2              1            -3   0  ...                  0          None   
3              1             0   0  ...                  0          None   
4              1             0   0  ...                  0          None   

   posteam_timeouts_pre HomeTimeouts_Remaining_Pre AwayTimeouts_Remaining_Pre  \
0                     3                          3                          3   
1                     3                          3                          3   
2                     3                          3                          3   
3                     3                          3                          3   
4                     3                          3                          3   

   HomeTimeouts_Remaining_Post  AwayTimeouts_Remaining_Post  ExPoint_Prob  \
0                            3                            3           0.0   
1                            3                            3           0.0   
2                            3                            3           0.0   
3                            3                            3           0.0   
4                            3                            3           0.0   

   TwoPoint_Prob  Season  
0            0.0    2009  
1            0.0    2009  
2            0.0    2009  
3            0.0    2009  
4            0.0    2009  

[5 rows x 41 columns]

In [7]:
#Lets confirm if really all na_values were dropped
dropped_data.isnull().sum()

Date                           0
GameID                         0
Drive                          0
qtr                            0
TimeUnder                      0
ydstogo                        0
ydsnet                         0
PlayAttempted                  0
Yards.Gained                   0
sp                             0
Touchdown                      0
Safety                         0
Onsidekick                     0
PlayType                       0
Passer_ID                      0
PassAttempt                    0
AirYards                       0
YardsAfterCatch                0
QBHit                          0
InterceptionThrown             0
Rusher_ID                      0
RushAttempt                    0
Receiver_ID                    0
Reception                      0
Fumble                         0
Sack                           0
Challenge.Replay               0
Accepted.Penalty               0
Penalty.Yards                  0
HomeTeam                       0
AwayTeam  

In [8]:
#Checking the data, we can see although all na_values were dropped successfully, 61 columns were lost in the process
print('Columns of original data: %d \n' % nfl_data.shape[1])
print('Columns with dropped data: %d \n' % dropped_data.shape[1])
print('Amount of lost data: %d' % (nfl_data.shape[1] - dropped_data.shape[1]))

Columns of original data: 102 

Columns with dropped data: 41 

Amount of lost data: 61


**Let's try to break down our data into a smaller subset for better usability**

In [9]:
#Get a small subset of the NFL dataset
subset_nfl_data = nfl_data.loc[:, 'EPA':'Season'].head()
subset_nfl_data

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474       NaN       NaN     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760       NaN       NaN     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796       NaN       NaN     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758       NaN       NaN    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295       NaN       NaN    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712       NaN       NaN    2009

In [10]:
#Let's attempt filling all the missing values with 0
subset_nfl_data.fillna(0)

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474  0.000000  0.000000     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  0.000000  0.000000     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796  0.000000  0.000000     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758  0.000000  0.000000    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.000000  0.000000    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712  0.000000  0.000000    2009

In [11]:
#Another method will be to fill all na_values with values that comes directly after it in the same column
#Fill any remaining na_value with 0
subset_nfl_data.fillna(method='bfill', axis=0).fillna(0)

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474 -1.068169  1.146076     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  3.318841 -5.031425     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796  0.000000  0.000000     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758 -0.032244  0.036899    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.106663 -0.156239    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712  0.000000  0.000000    2009

# **Let's attempt another dataset**

In [12]:
from learntools.core import binder
binder.bind(globals())
from learntools.data_cleaning.ex1 import *
print('Setup Complete')

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3347: DtypeWarning: Columns (22,32) have mixed types.Specify dtype option on import or set low_memory=False.
  if (await self.run_code(code, result,  async_=asy)):


Setup Complete


In [13]:
#read in the data and check the head
frisco_permits = pd.read_csv('../input/building-permit-applications-data/Building_Permits.csv')
frisco_permits.head()

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3156: DtypeWarning: Columns (22,32) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Ellis            St  ...                         3.0   
1       Geary            St  ...                         3.0   
2     Pacific            Av  ...                         1.0   
3     Pacific            Av  ...                         5.0   
4      Market            St  ...                         3.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   
3                         wood frame (5)                        5.0   
4                          constr type 3                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   
3                         wood frame (5)         NaN                 3.0   
4                                    NaN         NaN                 6.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0   
2                        Russian Hill  94109.0   
3                            Nob Hill  94109.0   
4                          Tenderloin  94102.0   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  
3   (37.79595867909168, -122.41557405519474)  1443574295566  
4   (37.78315261897309, -122.40950883997789)   144548169992  

[5 rows x 43 columns]

In [14]:
frisco_permits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198900 entries, 0 to 198899
Data columns (total 43 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   Permit Number                           198900 non-null  object 
 1   Permit Type                             198900 non-null  int64  
 2   Permit Type Definition                  198900 non-null  object 
 3   Permit Creation Date                    198900 non-null  object 
 4   Block                                   198900 non-null  object 
 5   Lot                                     198900 non-null  object 
 6   Street Number                           198900 non-null  int64  
 7   Street Number Suffix                    2216 non-null    object 
 8   Street Name                             198900 non-null  object 
 9   Street Suffix                           196132 non-null  object 
 10  Unit                                    2947

In [15]:
#set seed for reproducibility
np.random.seed(0)

In [16]:
#Amount of missing values
missing_values = frisco_permits.isnull().sum()
missing_values

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

In [17]:
#Percentage missing values
total_cells = np.product(frisco_permits.shape)
total_missing_values = missing_values.sum()

percent_missing = (total_missing_values/total_cells)*100
print('The percentage of missing values is %.2f'%percent_missing+ '%')

The percentage of missing values is 26.26%


**26.26% is more than a quarter of our data missing**

In [18]:
frisco_permits[['Street Number Suffix','Zipcode']].isnull().sum()

Street Number Suffix    196684
Zipcode                   1716
dtype: int64

In [19]:
#Lets try a rows missing values removal
frisco_na_rows_dropped = frisco_permits.dropna(axis=0)
frisco_na_rows_dropped

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location, Record ID]
Index: []

[0 rows x 43 columns]

**All the rows in our dataset have na_values hence the dropping of all the data**

In [20]:
#Lets try a columns missing values removal
frisco_na_cols_dropped = frisco_permits.dropna(axis=1)
frisco_na_cols_dropped

Permit Number  Permit Type            Permit Type Definition  \
0       201505065519            4                      sign - erect   
1       201604195146            4                      sign - erect   
2       201605278609            3  additions alterations or repairs   
3       201611072166            8            otc alterations permit   
4       201611283529            6                       demolitions   
...              ...          ...                               ...   
198895       M862628            8            otc alterations permit   
198896  201712055595            8            otc alterations permit   
198897       M863507            8            otc alterations permit   
198898       M863747            8            otc alterations permit   
198899       M864287            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Name  \
0                05/06/2015  0326   023            140       Ellis   
1                04/19/2016  0306   007            440       Geary   
2                05/27/2016  0595   203           1647     Pacific   
3                11/07/2016  0156   011           1230     Pacific   
4                11/28/2016  0342   001            950      Market   
...                     ...   ...   ...            ...         ...   
198895           12/05/2017  0113  017A           1228  Montgomery   
198896           12/05/2017  0271   014            580        Bush   
198897           12/06/2017  4318   019           1568     Indiana   
198898           12/06/2017  0298   029            795      Sutter   
198899           12/07/2017  0160   006            838     Pacific   

       Current Status Current Status Date  Filed Date      Record ID  
0             expired          12/21/2017  05/06/2015  1380611233945  
1              issued          08/03/2017  04/19/2016  1420164406718  
2           withdrawn          09/26/2017  05/27/2016  1424856504716  
3            complete          07/24/2017  11/07/2016  1443574295566  
4              issued          12/01/2017  11/28/2016   144548169992  
...               ...                 ...         ...            ...  
198895         issued          12/05/2017  12/05/2017  1489337276729  
198896         issued          12/06/2017  12/05/2017  1489462354993  
198897         issued          12/06/2017  12/06/2017  1489539379952  
198898         issued          12/06/2017  12/06/2017  1489608233656  
198899         issued          12/07/2017  12/07/2017  1489796283803  

[198900 rows x 12 columns]

In [21]:
frisco_na_cols_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198900 entries, 0 to 198899
Data columns (total 12 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Permit Number           198900 non-null  object
 1   Permit Type             198900 non-null  int64 
 2   Permit Type Definition  198900 non-null  object
 3   Permit Creation Date    198900 non-null  object
 4   Block                   198900 non-null  object
 5   Lot                     198900 non-null  object
 6   Street Number           198900 non-null  int64 
 7   Street Name             198900 non-null  object
 8   Current Status          198900 non-null  object
 9   Current Status Date     198900 non-null  object
 10  Filed Date              198900 non-null  object
 11  Record ID               198900 non-null  int64 
dtypes: int64(3), object(9)
memory usage: 18.2+ MB


In [22]:
print('We lost {} columns'.format(frisco_permits.shape[1] - frisco_na_cols_dropped.shape[1]))

We lost 31 columns


In [23]:
#Using bfill to fill in NaN values
frisco_permits_with_na_imputed = frisco_permits.fillna(method='bfill', axis=0).fillna(0)
frisco_permits_with_na_imputed

Permit Number  Permit Type            Permit Type Definition  \
0       201505065519            4                      sign - erect   
1       201604195146            4                      sign - erect   
2       201605278609            3  additions alterations or repairs   
3       201611072166            8            otc alterations permit   
4       201611283529            6                       demolitions   
...              ...          ...                               ...   
198895       M862628            8            otc alterations permit   
198896  201712055595            8            otc alterations permit   
198897       M863507            8            otc alterations permit   
198898       M863747            8            otc alterations permit   
198899       M864287            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Number Suffix  \
0                05/06/2015  0326   023            140                    A   
1                04/19/2016  0306   007            440                    A   
2                05/27/2016  0595   203           1647                    A   
3                11/07/2016  0156   011           1230                    A   
4                11/28/2016  0342   001            950                    A   
...                     ...   ...   ...            ...                  ...   
198895           12/05/2017  0113  017A           1228                    0   
198896           12/05/2017  0271   014            580                    0   
198897           12/06/2017  4318   019           1568                    0   
198898           12/06/2017  0298   029            795                    0   
198899           12/07/2017  0160   006            838                    0   

       Street Name Street Suffix  ...  Existing Construction Type  \
0            Ellis            St  ...                         3.0   
1            Geary            St  ...                         3.0   
2          Pacific            Av  ...                         1.0   
3          Pacific            Av  ...                         5.0   
4           Market            St  ...                         3.0   
...            ...           ...  ...                         ...   
198895  Montgomery            St  ...                         5.0   
198896        Bush            St  ...                         5.0   
198897     Indiana            St  ...                         0.0   
198898      Sutter            St  ...                         0.0   
198899     Pacific            Av  ...                         0.0   

       Existing Construction Type Description Proposed Construction Type  \
0                               constr type 3                        1.0   
1                               constr type 3                        1.0   
2                               constr type 1                        1.0   
3                              wood frame (5)                        5.0   
4                               constr type 3                        1.0   
...                                       ...                        ...   
198895                         wood frame (5)                        5.0   
198896                         wood frame (5)                        5.0   
198897                                      0                        0.0   
198898                                      0                        0.0   
198899                                      0                        0.0   

       Proposed Construction Type Description Site Permit Supervisor District  \
0                               constr type 1           Y                 3.0   
1                               constr type 1           Y                 3.0   
2                               constr type 1           Y                 3.0   
3                              wood frame (5)           Y                 3.0   
4                               constr type 1           Y 

In [24]:
frisco_permits_with_na_imputed.isnull().sum()

Permit Number                             0
Permit Type                               0
Permit Type Definition                    0
Permit Creation Date                      0
Block                                     0
Lot                                       0
Street Number                             0
Street Number Suffix                      0
Street Name                               0
Street Suffix                             0
Unit                                      0
Unit Suffix                               0
Description                               0
Current Status                            0
Current Status Date                       0
Filed Date                                0
Issued Date                               0
Completed Date                            0
First Construction Document Date          0
Structural Notification                   0
Number of Existing Stories                0
Number of Proposed Stories                0
Voluntary Soft-Story Retrofit   

* 